File: bwu/linked_list/92-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        start = dummy
        curr = dummy.next
        
        while curr and m > 1:
            curr = curr.next
            start = start.next
            m -= 1
            n -= 1
            
        while curr.next and n > 1:
            tmp = start.next
            new_next  = curr.next
            start.next = new_next
            curr.next = new_next.next
            new_next.next = tmp
            n -= 1
            
        return dummy.next

File: captainhcg/go/92-reverse-linked-list-ii.go
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	dummy := &ListNode{0, head}
	pre := dummy
	for i := 1; i < m; i += 1 {
		pre = pre.Next
	}
	start, end := pre.Next, pre.Next
	for i := 0; i < n-m; i += 1 {
		oldStart := start
		start, end.Next = end.Next, end.Next.Next
		start.Next, pre.Next = oldStart, start
	}
	return dummy.Next
}
